instrumentation/reactor-netty/reactor-netty-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactornetty/v1_0/ResponseReceiverInstrumentation.java

```diff
@@ -0,0 +1,237 @@
+
+  @Override
+  public ElementMatcher<ClassLoader> classLoaderOptimization() {
+    return hasClassesNamed("reactor.netty.http.client.HttpClient$ResponseReceiver");
+  }
+
+  @Override
+  public ElementMatcher<TypeDescription> typeMatcher() {
+    return implementsInterface(named("reactor.netty.http.client.HttpClient$ResponseReceiver"));
+  }
+
+  @Override
+  public void transform(TypeTransformer transformer) {
+    transformer.applyAdviceToMethod(
+        named("response").and(takesArguments(0)).and(returns(named("reactor.core.publisher.Mono"))),
+        this.getClass().getName() + "$ResponseMonoAdvice");
+    transformer.applyAdviceToMethod(
+        named("response")
+            .and(takesArguments(1))
+            .and(takesArgument(0, BiFunction.class))
+            .and(returns(named("reactor.core.publisher.Flux"))),
+        this.getClass().getName() + "$ResponseFluxAdvice");
+    transformer.applyAdviceToMethod(
+        named("responseConnection")
+            .and(takesArguments(1))
+            .and(takesArgument(0, BiFunction.class))
+            .and(returns(named("reactor.core.publisher.Flux"))),
+        this.getClass().getName() + "$ResponseConnectionAdvice");
+    transformer.applyAdviceToMethod(
+        named("responseContent")
+            .and(takesArguments(0))
+            .and(returns(named("reactor.netty.ByteBufFlux"))),
+        this.getClass().getName() + "$ResponseContentAdvice");
+    transformer.applyAdviceToMethod(
+        named("responseSingle")
+            .and(takesArguments(1))
+            .and(takesArgument(0, BiFunction.class))
+            .and(returns(named("reactor.core.publisher.Mono"))),
+        this.getClass().getName() + "$ResponseSingleAdvice");
+  }
+
+  @SuppressWarnings("unused")
+  public static class ResponseMonoAdvice {
+
+    @Advice.OnMethodEnter(suppress = Throwable.class, skipOn = Advice.OnNonDefaultValue.class)
+    public static HttpClient.ResponseReceiver<?> onEnter(
+        @Advice.Local("otelCallDepth") CallDepth callDepth,
+        @Advice.This HttpClient.ResponseReceiver<?> receiver) {
+
+      callDepth = CallDepth.forClass(ClassLoader.class);
+      if (callDepth.getAndIncrement() > 0) {
+        // execute the original method on nested calls
+        return null;
+      }
+
+      // non-null value will skip the original method invocation
+      return HttpResponseReceiverInstrumenter.instrument(receiver);
+    }
+
+    @Advice.OnMethodExit(suppress = Throwable.class)
+    public static void onExit(
+        @Advice.Local("otelCallDepth") CallDepth callDepth,
+        @Advice.Enter HttpClient.ResponseReceiver<?> modifiedReceiver,
+        @Advice.Return(readOnly = false) Mono<HttpClientResponse> returnValue) {
+
+      if (modifiedReceiver != null) {
+        returnValue = modifiedReceiver.response();
+      }
+
+      // needs to be called after original method to prevent StackOverflowError
+      callDepth.decrementAndGet();
```

Comment:
can you move the callDepth update to beginning of these methods? I'm just paranoid about resetting thread local state 😄

Reply:
I have to decrement the count _after_ the original `response*()` method is called, that's why they're the last thing I'm calling. I've wrapped those invocations in try-finally blocks - now the counter will always be decremented.

Reply:
oh, i completely missed the `skipOn` and what was happening here 👍